### PR TITLE
Tally sheet edit PE-22

### PR DESCRIPTION
**Pull request by callkalpa**
[#475] Tally sheet edit PE-39

### CODE REVIEW

File: rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/sample-config-data/1/invalid-vote-categories-dataset.csv [hunk 1]
```diff
@@ -1,8 +1,11 @@
-Invalid Vote Category Description
+Invalid Vote Category Description,Invalid Vote Category Type
 Does not bear the official mark.
 Votes are given for more than one recognized political party.
 Votes are given for more than one Independent Group.
```
Comment (dinukadesilva) [inline rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/sample-config-data/1/invalid-vote-categories-dataset.csv]: Shall we add a type to these two rather keeping them empty

File: rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/__init__.py [hunk 2]
```diff
@@ -293,6 +296,49 @@ def build_election(self, party_candidate_dataset_file=None,
             ]
         )
 
+        tally_sheet_template_pe_22 = Template.create(
+            templateName=PE_22
+        )
+        tally_sheet_template_pe_22_preferences_indicated_for_more_than_three_candidates_row = tally_sheet_template_pe_22.add_row(
```
Comment (dinukadesilva) [inline rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/__init__.py]: Since invalidVoteCategoryId is taken as a column,is it required to have multiple rows?

File: rest/results-tabulation-api/migrations/versions/5f61093c9695_.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+# revision identifiers, used by Alembic.
+revision = '5f61093c9695'
+down_revision = 'ab9d4f52637d'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    ### commands auto generated by Alembic - please adjust! ###
+    op.alter_column('election', 'isListed',
```
Comment (dinukadesilva) [inline rest/results-tabulation-api/migrations/versions/5f61093c9695_.py]: Due to a bug in the migration file generation, some of this existing changes are re populated. Shall we remove all unnecessary changes

File: rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/RoleBasedAccess.py [hunk 4]
```diff
@@ -22,6 +22,10 @@
             NonPostal: [READ, WRITE, SUBMIT, LOCK],
             Postal: [READ, WRITE, SUBMIT, LOCK]
         },
+        PE_22: {
```
Comment (dinukadesilva) [inline rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/RoleBasedAccess.py:25]: There's few places down to add few entries. Check how it's done for PE-4

**Review by dinukadesilva — CHANGES_REQUESTED** [review]

File: rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/sample-config-data/1/invalid-vote-categories-dataset.csv (hunk 1 above)
Comment (callkalpa) [reply rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/sample-config-data/1/invalid-vote-categories-dataset.csv]: Added a default type of "ELECTION"

File: rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/__init__.py (hunk 2 above)
Comment (callkalpa) [reply rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/__init__.py]: Fixed.

File: rest/results-tabulation-api/migrations/versions/5f61093c9695_.py (hunk 3 above)
Comment (callkalpa) [reply rest/results-tabulation-api/migrations/versions/5f61093c9695_.py]: Fixed.

File: rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/RoleBasedAccess.py (hunk 4 above)
Comment (callkalpa) [reply rest/results-tabulation-api/ext/ExtendedElection/ExtendedElectionParliamentaryElection2020/RoleBasedAccess.py:25]: Fixed.

**Comment by callkalpa** [comment]
@dinukadesilva I have fixed the issues.

File: rest/results-tabulation-api/migrations/versions/5f61093c9695_.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+# revision identifiers, used by Alembic.
+revision = '5f61093c9695'
+down_revision = 'ab9d4f52637d'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('election_invalidVoteCategory', sa.Column('invalidVoteCategoryType', sa.String(length=50), nullable=True))
```
Comment (dinukadesilva) [inline rest/results-tabulation-api/migrations/versions/5f61093c9695_.py]: Have to update the existing entries here.

**Review by dinukadesilva — CHANGES_REQUESTED** [review]

File: rest/results-tabulation-api/migrations/versions/5f61093c9695_.py (hunk 5 above)
Comment (dinukadesilva) [reply rest/results-tabulation-api/migrations/versions/5f61093c9695_.py]: https://github.com/ECLK/results-tabulation-api/blob/da46ab37c1d417efe661b838de3780ba37159562/rest/results-tabulation-api/migrations/versions/0718cd52821c_.py

@callkalpa You could refer this file which would help

File: web/results-tabulation-ui/src/components/election/extended-election/ParliamentElection2020/index.js [hunk 6]
```diff
@@ -45,11 +46,11 @@ export default class ExtendedElectionParliamentElection2020 extends ExtendedElec
                         let tallySheetCodes = [];
                         let tallySheetCodeLabels = [];
                         if (voteType === "Postal") {
-                            tallySheetCodes = [TALLY_SHEET_CODE_CE_201_PV, TALLY_SHEET_CODE_PE_27, TALLY_SHEET_CODE_PE_39];
-                            tallySheetCodeLabels = ["CE 201 PV (Postal)", "PE-27 PV (Postal)", "PE-39 (Postal)"];
+                            tallySheetCodes = [TALLY_SHEET_CODE_CE_201_PV, TALLY_SHEET_CODE_PE_27, TALLY_SHEET_CODE_PE_39, TALLY_SHEET_CODE_PE_22];
```
Comment (dinukadesilva) [inline web/results-tabulation-ui/src/components/election/extended-election/ParliamentElection2020/index.js]: PE-22 must be under preferences

**Review by dinukadesilva — CHANGES_REQUESTED** [review]

**Review by dinukadesilva — APPROVED** [review]
